pkg/genericapiserver/genericapiserver.go

```diff
@@ -0,0 +1,541 @@
+	ServiceClusterIPRange *net.IPNet
+
+	// The IP address for the Genericapiserver service (must be inside ServiceClusterIPRange
+	ServiceReadWriteIP net.IP
+
+	// The range of ports to be assigned to services with type=NodePort or greater
+	ServiceNodePortRange util.PortRange
+
+	// Used to customize default proxy dial/tls options
+	ProxyDialer          apiserver.ProxyDialerFunc
+	ProxyTLSClientConfig *tls.Config
+
+	// Additional ports to be exposed on the Genericapiserver service
+	// extraServicePorts is injectable in the event that more ports
+	// (other than the default 443/tcp) are exposed on the Genericapiserver
+	// and those ports need to be load balanced by the Genericapiserver
+	// service because this pkg is linked by out-of-tree projects
+	// like openshift which want to use the Genericapiserver but also do
+	// more stuff.
+	ExtraServicePorts []api.ServicePort
+	// Additional ports to be exposed on the Genericapiserver endpoints
+	// Port names should align with ports defined in ExtraServicePorts
+	ExtraEndpointPorts []api.EndpointPort
+
+	KubernetesServiceNodePort int
+}
+
+// Genericapiserver contains state for a Kubernetes cluster Genericapiserver/api server.
+type Genericapiserver struct {
+	// "Inputs", Copied from Config
+	ServiceClusterIPRange *net.IPNet
+	ServiceNodePortRange  util.PortRange
+	cacheTimeout          time.Duration
+	MinRequestTimeout     time.Duration
+
+	mux                      apiserver.Mux
+	MuxHelper                *apiserver.MuxHelper
+	HandlerContainer         *restful.Container
+	RootWebService           *restful.WebService
+	enableLogsSupport        bool
+	enableUISupport          bool
+	enableSwaggerSupport     bool
+	enableProfiling          bool
+	enableWatchCache         bool
+	ApiPrefix                string
+	ApiGroupPrefix           string
+	corsAllowedOriginList    []string
+	authenticator            authenticator.Request
+	authorizer               authorizer.Authorizer
+	AdmissionControl         admission.Interface
+	MasterCount              int
+	ApiGroupVersionOverrides map[string]APIGroupVersionOverride
+	RequestContextMapper     api.RequestContextMapper
+
+	// External host is the name that should be used in external (public internet) URLs for this Genericapiserver
+	externalHost string
+	// ClusterIP is the IP address of the Genericapiserver within the cluster.
+	ClusterIP            net.IP
+	PublicReadWritePort  int
+	ServiceReadWriteIP   net.IP
+	ServiceReadWritePort int
+	masterServices       *util.Runner
+	ExtraServicePorts    []api.ServicePort
+	ExtraEndpointPorts   []api.EndpointPort
+
+	// storage contains the RESTful endpoints exposed by this Genericapiserver
+	storage map[string]rest.Storage
+
+	// "Outputs"
+	Handler         http.Handler
+	InsecureHandler http.Handler
+
+	// Used for custom proxy dialing, and proxy TLS options
+	ProxyTransport http.RoundTripper
+
+	KubernetesServiceNodePort int
+}
+
+func (s *Genericapiserver) StorageDecorator() generic.StorageDecorator {
```

Comment:
I moved it to server since I wanted to call it from a method in master that did not have access to the config.
Eventually, I ended up passing config to that function for something else :) So I can move this back to config, if you think its better?